lib/remap-functions/src/encode_base64.rs

```diff
@@ -0,0 +1,143 @@
+use remap::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct EncodeBase64;
+
+impl Function for EncodeBase64 {
+    fn identifier(&self) -> &'static str {
+        "encode_base64"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "padding",
+                accepts: |v| matches!(v, Value::Boolean(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let padding = arguments.optional("padding").map(Expr::boxed);
+
+        Ok(Box::new(EncodeBase64Fn { value, padding }))
+    }
+}
+
+#[derive(Clone, Debug)]
+struct EncodeBase64Fn {
+    value: Box<dyn Expression>,
+    padding: Option<Box<dyn Expression>>,
+}
+
+impl Expression for EncodeBase64Fn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let value = self.value.execute(state, object)?.try_bytes()?;
+
+        let padding = match &self.padding {
```

Comment:
Nit: could we use `map` and `unwrap_or` here?

```rust
 let padding = self.padding.map(|p| p.execute(state, object)?.try_boolean()?)).unwrap_or(true);
```

This would let you end up with just a boolean rather than an `Option`, simplifying the below handling.

Reply:
This specific expression doesn't quite work but I'll see if I can put together a comparable one-liner

Reply:
The specific issue is that you unfortunately can't casually pepper `map` statements with `?` the way that I suspect we both wish you could 😄 I played with this a bit and I have yet to come up with anything more clear than this.

Reply:
Something like this should work:

```rust
let padding = padding
	.map(|p| p.execute(state, object).and_then(Value::try_boolean)) // Option<Result<bool>>
	.transpose()? // Option<bool>
	.unwrap_or(true); // bool
```

Or, combined with the next part:

```rust
let config = padding
	.map(|p| p.execute(state, object).and_then(Value::try_boolean)) // Option<Result<bool>>
	.transpose()? // Option<bool>
	.filter(|padding| padding) // Option<bool>
	.map(|_| base64::STANDARD) // Option<base64>
	.unwrap_or(base64::STANDARD_NO_PAD); // base64
```

Reply:
Unfortunately that also doesn't work, as the error types inside the map operation are mismatched (`remap_lang::Error` vs. `remap_lang::value::Error`). That's the issue I initially ran into and couldn't see beyond. I'll see what I can do.

Reply:
Ah right. You can call `into` on the error to convert it to a remap::Error

Reply:
I wasn't able to formulate a one-liner for this that provides the expected behavior. If someone else has an idea, please feel free to push to this PR.

Reply:
:+1: I pushed ac0fa6e534388cb6b09570e14f707fb31bd3fe93 with my vision for it :smile: